cmaes.lua

```diff
@@ -0,0 +1,273 @@
+      local y = xmean - xold
+      local z = invsqrtC * y -- == C^(-1/2) * (xnew - xold)
+
+      local c = (cs * (2-cs) * mueff)^0.5 / sigma
+      ps = ps - ps * cs + z * c -- exponential decay on ps
+      local hsig = (torch.sum(torch.pow(ps,2)) / 
+        (1-(1-cs)^(2*counteval/lambda)) / N  < 2 + 4./(N+1))
+      hsig = hsig and 1.0 or 0.0 --use binary numbers
+
+      local c = (cc * (2-cc) * mueff)^0.5 / sigma
+      pc = pc - pc * cc + y * c * hsig -- exponential decay on pc
+
+      -- Adapt covariance matrix C
+      local c1a = c1 - (1-hsig^2) * c1 * cc * (2-cc)
+      -- for a minor adjustment to the variance loss by hsig
+      for i=1,N do
+        for j=1,N do
+          local r = torch.range(1,mu)
+          r:apply(function(k) 
+              return weights[k] * (arx[k][i]-xold[i]) * (arx[k][j]-xold[j]) end)
+          Cmuij = torch.sum(r) / sigma^2  -- rank-mu update
+          C[i][j] = C[i][j] + ((-c1a - cmu) * C[i][j] + 
+              c1 * pc[i]*pc[j] + cmu * Cmuij)
+          end
+        end
+
+        -- Adapt step-size sigma with factor <= exp(0.6) \approx 1.82
+        sigma = sigma * math.exp(math.min(0.6, 
+            (cs / damps) * (torch.sum(torch.pow(ps,2))/N - 1)/2))
+
+      end
+
+      local function stop() 
+        --[[return satisfied termination conditions in a table like 
+        {'termination reason':value, ...}, for example {'tolfun':1e-12}, 
+        or the empty table {}--]] 
+        res = {}
+        if counteval > 0 then
+          if counteval >= maxEval then
+            res['evals'] = maxEval
+          end
+          if ftarget ~= nil and fitvals:nElement() > 0 and fitvals[1] <= ftarget then
+            res['ftarget'] = ftarget
+          end
+          if torch.max(D) > 1e7 * torch.min(D) then
+            res['condition'] = 1e7
+          end
+          if fitvals:nElement() > 1 and fitvals[fitvals:nElement()] - fitvals[1] < 1e-12 then
+            res['tolfun'] = 1e-12 
+          end
+          if sigma * torch.max(D) < 1e-11 then
+            -- remark: max(D) >= max(diag(C))^0.5
+            res['tolx'] = 1e-11
+          end
+        end
+        return res
+      end
+
+      local function disp(verb_modulo)
+        --[[display some iteration info--]]
+        if verb_disp == 0 then
+          return nil
+        end
+        local iteration = counteval / lambda
+
+        if iteration == 1 or iteration % (10*verb_modulo) == 0 then
+          print('evals:\t ax-ratio max(std)   f-value')
+        end
+        if iteration <= 2 or iteration % verb_modulo == 0 then
+          local max_std = math.sqrt(torch.max(torch.diag(C)))
+          print(tostring(counteval).. ': ' .. 
+            string.format(' %6.1f %8.1e ', torch.max(D) / torch.min(D), sigma * max_std) 
+            .. tostring(fitvals[1]))
+        end
+
+        return nil
+      end
+
+      while next(stop()) == nil or iteration < min_iterations do
+        if iterations and iteration >= iterations then
```

Comment:
Actually iterations was a relict from an earlier version, I removed it entirely.